File: packages/tools/src/lib/tool-verification.ts
import { Image } from './image';
import { Video } from './video';

export interface ToolStatus {
    available: boolean;
    version?: string;
    error?: string;
}

export interface ToolsStatus {
    magick: ToolStatus;
    ffprobe: ToolStatus;
    ffmpeg: ToolStatus;
    allAvailable: boolean;
    missingTools: string[];
}

/**
 * Check the availability of all required tools
 */
export async function verifyTools(): Promise<ToolsStatus> {
    const [magickStatus, ffprobeStatus, ffmpegStatus] = await Promise.all([
        Image.verifyImageMagick(),
        Video.verifyFfprobe(),
        Video.verifyFfmpeg()
    ]);

    const missingTools: string[] = [];
    
    if (!magickStatus.available) {
        missingTools.push('ImageMagick');
    }
    if (!ffprobeStatus.available) {
        missingTools.push('ffprobe');
    }
    if (!ffmpegStatus.available) {
        missingTools.push('ffmpeg');
    }

    return {
        magick: magickStatus,
        ffprobe: ffprobeStatus,
        ffmpeg: ffmpegStatus,
        allAvailable: missingTools.length === 0,
        missingTools
    };
}

/**
 * Check if all required tools are available
 */
export async function ensureToolsAvailable(options?: {
    promptForInstall?: boolean;
    silent?: boolean;
}): Promise<boolean> {
    const { silent = false } = options || {};
    
    const toolsStatus = await verifyTools();
    
    if (toolsStatus.allAvailable) {
        if (!silent) {
            console.log('✓ All required tools are available:');
            if (toolsStatus.magick.version) {
                console.log(`  • ImageMagick v${toolsStatus.magick.version}`);
            }
            if (toolsStatus.ffprobe.version) {
                console.log(`  • ffprobe v${toolsStatus.ffprobe.version}`);
            }
            if (toolsStatus.ffmpeg.version) {
                console.log(`  • ffmpeg v${toolsStatus.ffmpeg.version}`);
            }
        }
        return true;
    }

    return false;
}

File: packages-zig/tools-zig/src/lib/tool-verification.zig
const std = @import("std");
const Image = @import("image.zig").Image;
const Video = @import("video.zig").Video;

//
// The availability of one tool.
//
pub const ToolStatus = struct {
    // True when the tool can be run.
    available: bool,

    // The tool version, when available.
    version: ?[]const u8 = null,

    // Why the tool is not available.
    @"error": ?[]const u8 = null,
};

//
// The availability of all the tools Photosphere needs.
//
pub const ToolsStatus = struct {
    // ImageMagick.
    magick: ToolStatus,

    // ffprobe.
    ffprobe: ToolStatus,

    // ffmpeg.
    ffmpeg: ToolStatus,

    // True when every tool is available.
    allAvailable: bool,

    // Display names of the tools that are not available.
    missingTools: []const []const u8,
};

//
// Check the availability of all required tools
//
pub fn verifyTools(allocator: std.mem.Allocator, io: std.Io) !ToolsStatus {
    // Serialize tool verification to avoid race conditions
    const magickStatus = try Image.verifyImageMagick(allocator, io);
    const ffprobeStatus = Video.verifyFfprobe(allocator, io);
    const ffmpegStatus = Video.verifyFfmpeg(allocator, io);

    var missingTools: std.ArrayList([]const u8) = .empty;

    if (!magickStatus.available) {
        try missingTools.append(allocator, "ImageMagick");
    }
    if (!ffprobeStatus.available) {
        try missingTools.append(allocator, "ffprobe");
    }
    if (!ffmpegStatus.available) {
        try missingTools.append(allocator, "ffmpeg");
    }

    return .{
        .magick = .{ .available = magickStatus.available, .version = magickStatus.version, .@"error" = magickStatus.@"error" },
        .ffprobe = .{ .available = ffprobeStatus.available, .version = ffprobeStatus.version, .@"error" = ffprobeStatus.@"error" },
        .ffmpeg = .{ .available = ffmpegStatus.available, .version = ffmpegStatus.version, .@"error" = ffmpegStatus.@"error" },
        .allAvailable = missingTools.items.len == 0,
        .missingTools = missingTools.items,
    };
}

// Not ported: ensureToolsAvailable (not used by replicate or verify).
